lib/Sema/ConstraintSystem.cpp

```diff
@@ -1589,6 +1684,20 @@ Type ConstraintSystem::getUnopenedTypeOfReference(
   return requestedType;
 }
 
+void ConstraintSystem::recordOpenedType(
+    ConstraintLocator *locator, ArrayRef<OpenedType> openedTypes) {
+  bool inserted = OpenedTypes.insert({locator, openedTypes}).second;
+  if (inserted) {
```

Comment:
I think every spot where this is conditional needs to become an assert, attempting to re-record the same thing multiple times (especially “opened” type variables) might be a symptom of larger issues…

Reply:
They’re all coming from applySolution() with partial solutions I believe.

Reply:
My plan there is to replace partial solutions with a “redo” of a trail segment

Reply:
I know that we currently do that but since you are refactoring this I think we should try to make it an assert even if we have to add logic to reconcile data in `applySolution`.

Reply:
Re: plan, sounds good!

Reply:
The latest PR addressed this